ansible_base/authentication/authenticator_plugins/radius.py

```diff
@@ -0,0 +1,323 @@
+        label=_("RADIUS Secret"),
+        help_text=_("Shared secret for authenticating to RADIUS server."),
+        ui_field_label=_("Shared secret for authenticating to RADIUS server."),
+        required=True,
+    )
+
+
+class AuthenticatorPlugin(AbstractAuthenticatorPlugin):
+    configuration_class = RADIUSConfiguration
+    type = "RADIUS"
+    category = "password"
+    configuration_encrypted_fields = ["SECRET"]
+
+    def __init__(self, database_instance=None, *args, **kwargs):
+        super().__init__(database_instance, *args, **kwargs)
+        self.set_logger(logger)
+
+    def authenticate(self, request, username, password, **kwargs):
+        settings = SimpleNamespace(
+            RADIUS_SERVER=self.settings["HOST"],
+            RADIUS_PORT=self.settings["PORT"],
+            RADIUS_SECRET=self.settings["SECRET"],
+        )
+        backend = RADIUSBackend(settings)
+        user = backend.authenticate(request, username, password)
+        if user is None:
+            return None
+
+        get_or_create_authenticator_user(
+            user_id=username,
+            user_details={"username": username},
+            authenticator=self.database_instance,
+            extra_data={},
+        )
+        return user
+
+
+# The code below is based on the source code of the django-radius library (https://github.com/robgolding/django-radius).
+
+# Copyright (c) 2015, Rob Golding. All rights reserved.
+#
+# Redistribution and use in source and binary forms, with or without
+# modification, are permitted provided that the following conditions are met:
+#
+#     * Redistributions of source code must retain the above copyright
+#       notice, this list of conditions and the following disclaimer.
+#     * Redistributions in binary form must reproduce the above copyright
+#       notice, this list of conditions and the following disclaimer in the
+#       documentation and/or other materials provided with the distribution.
+#     * Neither the name of Rob Golding, nor the names of its contributors may
+#       be used to endorse or promote products derived from this software
+#       without specific prior written permission.
+#
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
+# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
+# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+DICTIONARY = u"""
```

Comment:
I agree, I can make it multiple comments first importing original implementation and then adjusting it to our needs, so that we don't loose history.
Actually I would like to move the derived code to a separate file, so that it's easier to maintain those changes and separate the 3rd party code from our code. Can you please suggest where is the best place within our project structure to put this file?

Reply:
I'm not sure it matters all to much at this time since we want to (eventually) split these out into their own repo. That being said, for the GitHub authenticators @jctanner needed a common configuration that he put in `authentication/authenticator_configurations`.  Since this is radius specific I might add something like `authentication/radius_backend` or something along those lines if you are going to have multiple files. If you don't want/need a full directory structure you could also put a single file into `utils`.